src/main/scala/util/RunUtils.scala

```diff
@@ -393,134 +623,76 @@ object RunUtils {
-                    val newLabel: String = block.label + t.name
-                    val directCall = DirectCall(t, indirectCall.returnTarget)
-                    newBlocks.append(Block(newLabel, None, ArrayBuffer(assume), directCall))
-                  }
-                  procedure.addBlocks(newBlocks)
-                  val newCall = GoTo(newBlocks, indirectCall.label)
-                  block.replaceJump(newCall)
-                }
-              case LiftedBottom =>
-              }
-          case _ =>
-      case _ =>
-    }
+    IRTransform.doCleanup(ctx)
 
-    def nameExists(name: String): Boolean = {
-      IRProgram.procedures.exists(_.name.equals(name))
-    }
+    q.loading.dumpIL.foreach(s => writeToFile(serialiseIL(ctx.program), s"$s-before-analysis.il"))
+    val analysis = q.staticAnalysis.map(conf => staticAnalysis(conf, ctx))
+    q.loading.dumpIL.foreach(s => writeToFile(serialiseIL(ctx.program), s"$s-after-analysis.il"))
 
-    def addFakeProcedure(name: String): Unit = {
-      IRProgram.procedures += Procedure(name)
+    if (q.runInterpret) {
+      val interpreter = Interpreter()
+      interpreter.interpret(ctx.program)
     }
 
-    def resolveAddresses(valueSet: Set[Value]): Set[AddressValue] = {
-      var functionNames: Set[AddressValue] = Set()
-      valueSet.foreach {
-        case globalAddress: GlobalAddress =>
-          if (nameExists(globalAddress.name)) {
-            functionNames += globalAddress
-            Logger.info(s"RESOLVED: Call to Global address ${globalAddress.name} resolved.")
-          } else {
-            addFakeProcedure(globalAddress.name)
-            functionNames += globalAddress
-            Logger.info(s"Global address ${globalAddress.name} does not exist in the program.  Added a fake function.")
-          }
-        case localAddress: LocalAddress =>
-          if (nameExists(localAddress.name)) {
-            functionNames += localAddress
-            Logger.info(s"RESOLVED: Call to Local address ${localAddress.name}")
-          } else {
-            addFakeProcedure(localAddress.name)
-            functionNames += localAddress
-            Logger.info(s"Local address ${localAddress.name} does not exist in the program. Added a fake function.")
-          }
-        case _ =>
+    IRTransform.prepareForTranslation(q.loading, ctx)
+
+    Logger.info("[!] Translating to Boogie")
+    val boogieTranslator = IRToBoogie(ctx.program, ctx.specification)
+    val boogieProgram = boogieTranslator.translate(q.boogieTranslation)
+
+    BasilResult(ctx, analysis, boogieProgram)
+  }
+
+  /** Use static analysis to resolve indirect calls and replace them in the IR until fixed point.
+    */
+  def staticAnalysis(config: StaticAnalysisConfig, ctx: IRContext): StaticAnalysisContext = {
+    var iteration = 1
+    var modified: Boolean = true
+    var analysisResult = mutable.ArrayBuffer[StaticAnalysisContext]()
```

Comment:
This doesn't need to be a var

Reply:
This function isn't a recursion anymore, this is incremented in a while loop so has to be var afaik?

Reply:
the analysisResult doesn't need to be a var is what I wanted to highlight